### DISCUSSION TITLE
Write playlist log to file

### DISCUSSION THREAD

**Discussion by muzo178** [savonet/liquidsoap#3777] [User Support]
store_metadata and on_metadata seems to be gone from the new releases.

I used to use something like this on 1.4.x:

```
input = store_metadata(size=10,id="history",input)
input = on_metadata(fun(meta) -> system("echo #{string.quote(meta['on_air'])}  #{string.quote(meta['artist'])} - #{string.quote(meta['title'])} >> /mnt/NAS/playlisthistory.txt"),input)
```

which use to create a text file like this:

```
2019/11/19 10:48:54 Hubert Kah - No Rain
2019/11/19 10:55:24 Joseph Malik - Evil Things (Bobby Hughes Combination Remix)
2019/11/19 11:02:47 Cool Seagull - Invisible Machine (Delta Wing Mix)
2019/11/19 11:07:45 Nazz - Loosen Up
2019/11/19 11:16:01 Hungry Lucy - Alfred (Haunted)
```

Any pointers as to how I can replicate the same functionality in 2.2.4?

**Comment by mooseh** [comment]
I beleive its source.on_metadata now so
as for the store_metadata, i don't think this will be needed, that's just to unsure it keeps a history of the last x metadata

```
input = source.on_metadata(fun(meta) -> system("echo #{string.quote(meta['on_air'])}  #{string.quote(meta['artist'])} - #{string.quote(meta['title'])} >> /mnt/NAS/playlisthistory.txt"),input)
```

**Comment by vitoyucepi** [comment]
Hi @muzo178,

If you're trying to write metadata to a file, you can use the [file.write](https://www.liquidsoap.info/doc-2.2.4/reference.html#file.write) function.
Each `source` should support an `on_metadata` method, which will trigger a callback function on new metadata being available.

```ruby
s = playlist("/tmp/test/music", mode="normal")

def log_metadata(m)
    on_air = m["on_air"]
    artist = m["artist"]
    title = m["title"]
    data = "#{on_air}  #{artist} - #{title}\n"
    file.write("/tmp/test/play.log", append=true, data=data)
end

s.on_metadata(log_metadata)
output.dummy(s, fallible=true)
```
